### DISCUSSION TITLE
TypeScript definitions for Tailwind CSS classes

### DISCUSSION THREAD

**Discussion by unrevised6419** [tailwindlabs/tailwindcss#5095] [Ideas]
Hello folks! I'm wondering, maybe already exists such a tool that can generate TypeScript types for tailwind classes. e.g.

```ts
export type BackgroundColor = 'bg-transparent' | 'bg-current' | 'bg-black' | 'bg-white' | 'bg-gray-50' // ...
export type TextColor = 'text-transparent' | 'text-current' | 'text-black' | 'text-white' | 'text-gray-50' // ...
```

I'm trying to write a basic script that generates these types, and wanted to know if I'm not doing this in vain, maybe exists such a generator.

This is a good helper if for example you want to type a component with direct tailwind classes, and want full TypeScript validation.

```tsx
import {BackgroundColor, TextColor} from "./tailwind.types";

type ButtonProps = {
    textColor: TextColor
    backgroundColor: BackgroundColor
}

function Button(props: PropsWithChildren<ButtonProps>) {
    return (
        <button className={clsx(props.textColor, props.backgroundColor, 'tw-p-3')}>
            {props.children}
        </button>
    )
}
```
```tsx
console.log(
    <Button textColor="text-blue-100" backgroundColor="bg-blue-300">
        Cool Button
    </Button>
)
```

Any advices are welcome. Maybe I can take some already utilities from tailwindcss that can make this generation process easier.

**Generate Types Script:** https://github.com/iamandrewluca/tailwind-types/blob/main/generate.mjs
**Generated Types:** https://github.com/iamandrewluca/tailwind-types/blob/main/tailwind.types.ts
**Example:** https://github.com/iamandrewluca/tailwind-types/blob/main/tailwind.example.tsx

The reason for this, is that at the moment I have to create a mapper object, that maps to a variant to a direct class, and I see it as a redundant layer of abstraction e.g.

```tsx
type TextColor = keyof typeof textColors
const textColors = {
  red: 'text-red-500'
}

type Props = {
  textColor: TextColor
}
```

And with time this mapper starts to grow, by adding almost all tailwind classes there.
Removing this layer of abstraction would be pretty cool.

Thanks!

**Comment by bennettdams** [comment]
@RobinMalfait Any idea if this is ever going to be implemented? 

I guess the VS Code plugin's language server/service already does something like this, otherwise IntelliSense wouldn't be able to make suggestions. Now the result of the language processing just has to be _transformed_ into TypeScript types. 

I would really love to have reusable classes, e.g. when writing different variants of reusable components. Something like:

```ts
const baseStyle: TailwindClass = 'leading-none rounded-full'

function Button() {
  ...
}
function ButtonSecondary() {
  ...
}
```

Wouldn't this also greatly improve the usability of Headless UI? Right now, fields on e.g. the `Transition` (e.g. `enterFrom="opacity-0"`) are never checked for being a valid Tailwind class.
